spec/remote_resource_spec.rb

```diff
@@ -0,0 +1,511 @@
+      end
+    end
+
+    context 'when Rails' do
+      before do
+        stub_const('Rails', true)
+      end
+
+      it 'forwards headers to the remote endpoint' do
+        headers = { 'Some-Rails' => 'header' }
+        ctx = double(request: double(env: {}, headers: double(to_h: headers)))
+        assert_headers(headers, ctx)
+      end
+
+      it 'still sends the JSONAPI Content-Type header' do
+        headers = { 'Content-Type' => 'application/vnd.api+json' }
+        ctx = double(request: double(env: {}, headers: double(to_h: headers)))
+        assert_headers(headers, ctx)
+      end
+
+      it 'can still override headers' do
+        klass.class_eval do
+          def request_headers
+            { 'Some-Foo' => 'bar' }
+          end
+        end
+        headers = { 'Some-Foo' => 'bar' }
+        ctx = double(request: double(env: {}, headers: double(to_h: headers)))
+        assert_headers(headers, ctx)
+      end
+    end
+
+    context 'when passed sorts' do
+      before do
+        params[:sort] = '-first_name,last_name'
+      end
+
+      it 'queries correctly' do
+        assert_params('sort=-first_name,last_name')
+      end
+    end
+
+    context 'when passed filters' do
+      before do
+        params[:filter] = { first_name: { suffix: 'foo' }, age: { gt: 1 } }
+      end
+
+      it 'queries correctly' do
+        assert_params('filter[age][gt]=1&filter[first_name][suffix]=foo')
+      end
+    end
+
+    context 'when passed pagination' do
+      before do
+        params[:page] = { size: 10, number: 2 }
+      end
+
+      it 'queries correctly' do
+        assert_params('page[number]=2&page[size]=10')
+      end
+    end
+
+    context 'when passed fields' do
+      before do
+        params[:fields] = { employees: 'first_name,age' }
+      end
+
+      it 'queries correctly' do
+        assert_params('fields[employees]=first_name,age')
+      end
+    end
+
+    context 'when passed extra fields' do
+      before do
+        params[:extra_fields] = { employees: 'foo,bar' }
+      end
+
+      it 'queries correctly' do
+        assert_params('extra_fields[employees]=foo,bar')
+      end
+    end
+
+    context 'when passed statistic' do
+      before do
+        params[:stats] = { total: 'count' }
+      end
+
+      it 'queries correctly' do
+        assert_params('stats[total]=count')
+      end
+    end
+
+    context 'when passed includes' do
+      before do
+        params[:include] = 'positions.department,teams'
+      end
+
+      context 'and the association is defined locally' do
+        before do
+          allow(PORO::PositionResource).to receive(:_all) { [] }
+          klass.class_eval do
+            has_many :positions, resource: PORO::PositionResource
+          end
+        end
+
+        it 'does not pass to the remote api' do
+          assert_params('include=teams')
+        end
+      end
+
+      context 'and the association is NOT defined locally' do
```

Comment:
Thought of one more branch in this condition.  If I have a remote resource, I can define its attributes, serializers, sideloads, etc, but only for the "root" remote resource.  If I explicitly define a sideload and that resource is also a remote resource for the same API, it will do two requests, yes?  Ideally we'd be able to make this do a single request but still leverage the local resource definition logic for serialization and such.

Reply:
I added a test for this. The way it works now is if the association is defined locally, but the associated is a remote resource with the same `remote_base_url`, it will be one request to the remote instead of two.